lib/anno_utils.pm

```diff
@@ -0,0 +1,2917 @@
+            croak "**rast_genome ERROR: GFF file is empty!\n";
+        }
+        ($gff_contents, $attr_delimiter) = $self->_parse_gff(
+											$gff_filename, $attr_delimiter);
+    }
+	return $gff_contents;
+}
+
+
+sub _get_contigs_from_fastafile {
+    my ($self, $fpath) = @_;
+
+    my @contigs;
+    my $contigID_hash = {};
+    my $fasta = "";
+    my $fh = $self->_openRead($fpath);
+    while (my $line = <$fh>) {
+        $fasta .= $line;
+    }
+    close($fh);
+    $fasta =~ s/\>([^\n]+)\n/>$1\|\|\|/g;
+    $fasta =~ s/\n//g;
+    my $array = [split(/\>/,$fasta)];
+    for (my $i=0; $i < @{$array}; $i++) {
+        if (scalar @contigs > $self->{max_contigs}) {
+            Bio::KBase::Exceptions::ArgumentValidationError->throw(
+                error => 'too many contigs',
+                method_name => 'anno_utils._get_contigs_from_fastafile'
+            );
+        }
+        if (length($array->[$i]) > 0) {
+            my $subarray = [split(/\|\|\|/,$array->[$i])];
+            if (@{$subarray} == 2) {
+                my $description = "unknown";
+                my $id = $subarray->[0];
+                if( $subarray->[0] =~ /^([^\s]+)\s(.+)$/) {
+                    $id = $1;
+                    $description = $2;
+                }
+                my $tmp_contigID = 'contigID_'.$i;
+                $contigID_hash->{$tmp_contigID} = $id;
+                my $contigobject = {
+                    id => $tmp_contigID,
+                    name => $tmp_contigID,
+                    "length" => length($subarray->[1]),
+                    md5 => Digest::MD5::md5_hex($subarray->[1]),
+                    sequence => $subarray->[1],
+                    description => $description
+                };
+                push(@contigs, $contigobject);
+            }
+        }
+    }
+    return (\@contigs, $contigID_hash);
+}
+
+
+#
+## get the contigs via WorkspaceClient and AssemblyUtilClient
+## Note: the input $ref has a value which is a Workspace.ref_string
+#
+sub _get_contigs {
+    my ($self, $ref) = @_;
+
+    return {} unless ($self->_validate_KB_objref($ref));
+    my $info = $self->_fetch_object_info($ref);
+    my ($contigs, $contigID_hash);
+
+    my $obj = {
+        _reference => $ref,
+        id => $info->[0],
+        name => $info->[1],
+        source_id => $info->[0],
+        source => "KBase",
+        type => "SingleGenome",
+        contigs => []
+    };
+    if ($info->[2] =~ /Assembly/) {
+        my $fpath = $self->_get_fasta_from_assembly($ref);
+        ($contigs, $contigID_hash) = $self->_get_contigs_from_fastafile($fpath);
+        $obj->{contigs} = $contigs;
+
+        my $sortedarray = [sort { $a->{sequence} cmp $b->{sequence} } @{$obj->{contigs}}];
+        my $str = "";
+        for (my $i=0; $i < @{$sortedarray}; $i++) {
+            if (length($str) > 0) {
+                $str .= ";";
+            }
+            $str .= $sortedarray->[$i]->{sequence};
+        }
```

Comment:
Don't use an iterator unless you need one.
Don't use an arrayref unless you need one.

You can rewrite this whole block in a more idiomatic perl manner like this:
```perl
my $str = join ";", sort { $a cmp $b } map { $_->{ sequence } } @$contigs;
```